MySQL: Create Table

jburchard opened the pull request:
Adds the `MySQL.copy()` method similar to `Redshift` and `Postgres` in which the table is automatically created in the database if it does not already exist.

Of note, the copy method utilizes the MySQL multi-insert functionality rather than `LOAD DATA`. While this introduces a slight performance hit (less so than other DBs), there are security concerns with `LOAD DATA LOCAL`. Additionally, many configurations of MySQL DBs don't allow `LOAD DATA` by default.

SorenSpicknall commented on parsons/databases/mysql/create_table.py:
```diff
@@ -0,0 +1,159 @@
+        # Convert to string to reevaluate data type
+        try:
+            t = ast.literal_eval(str(val))
+        except ValueError:
```
Can you outline the conditions for these two errors? Is one covering the case where the incoming value is already a string, and the other covering the case where that string is empty?

SorenSpicknall commented on parsons/databases/mysql/create_table.py:
```diff
@@ -0,0 +1,159 @@
+            if not self.is_valid_integer(val):
+                return 'varchar'
+            # Use smallest possible int type above TINYINT
+            if (-32768 < t < 32767) and current_type not in ['int', 'bigint', 'mediumint']:
```
Is there a system function call we could draw these boundary values from, rather than hardcoding them?

SorenSpicknall commented on parsons/databases/mysql/create_table.py:
```diff
@@ -0,0 +1,159 @@
+            # Use smallest possible int type above TINYINT
+            if (-32768 < t < 32767) and current_type not in ['int', 'bigint', 'mediumint']:
+                return 'smallint'
+            elif (-8388608 < t < 8388607) and current_type not in ['int', 'bigint']:
```
Is there a system function call we could draw these boundary values from, rather than hardcoding them?

SorenSpicknall commented on parsons/databases/mysql/create_table.py:
```diff
@@ -0,0 +1,159 @@
+                return 'smallint'
+            elif (-8388608 < t < 8388607) and current_type not in ['int', 'bigint']:
+                return 'mediumint'
+            elif (-2147483648 < t < 2147483647) and current_type not in ['bigint']:
```
Is there a system function call we could draw these boundary values from, rather than hardcoding them?

SorenSpicknall commented on parsons/databases/mysql/create_table.py:
```diff
@@ -0,0 +1,159 @@
+                if row_width > col_width:
+                    col_width = row_width
+
+        return col_type, col_width
```
Maybe we should default to making the column width wider than the existing maximum for rows already in place?

SorenSpicknall commented on parsons/databases/mysql/create_table.py:
```diff
@@ -0,0 +1,159 @@
+            col = col.strip()
+
+            # Replace spaces with underscores
+            col = col.replace(' ', '_')
```
Are there other special characters that need to be stripped out or replaced?

jburchard replied:
These were lifted from the `Redshift.copy()` and it has served as well to date. But, as corner cases arise, we can just slot them in.

jburchard commented:
@SorenSpicknall thanks for the feedback.

- Added config to the top of `create_table.py` that lifts the integer values out of the underlying code. I also included a link to the MySQL documentation around `INT` column type.

- Added a config at the top to set padding of varchar columns. The current value is `.25` which creates a 25% pad to the width. Does that seem like a good starting point?

Ready for rereview.
